### PR TITLE
docs(data): add document about how customize a entityCollection level property of ngrx-data

### PR DESCRIPTION
**Pull request by JiaLiPassion**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```
Add documentation about how to customize a EntityCollection level property. Such as in my use case, I want to add `paging information` in `EntityCollection level`, but current `document` can only add an additional property in `initialization`, there is no document about who to get the property from backend.

### CODE REVIEW

**Comment by brandonroberts** [comment]
Thanks for the PR @JiaLiPassion! Something is being formatted incorrectly in the preview starting at bullet 3 that throws off the rest of the sections.

**Comment by JiaLiPassion** [comment]
@brandonroberts, thanks for the review, I will fix it.

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 1]
```diff
@@ -220,6 +220,95 @@ The property values become the initial collection values for those properties wh
 
 The NgRx Data library generates selectors for these properties but has no way to update them. You'll have to create or extend the existing reducers to do that yourself.
 
+If the property you want to add comes from `backend`, you will need some additional work to make sure the property can be saved into store from `Effects` correctly.
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
If the property you want to add comes from `backend`, you will need some additional work to make sure the property can be saved into the store from `Effects` correctly.
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 2]
```diff
@@ -220,6 +220,95 @@ The property values become the initial collection values for those properties wh
...
+
+#### Step1: Implement `PersistenceResultHandler` to save data from backend to action.payload
+
+You need to create a `AdditioanlPersistenceResultHandler extends DefaultPersistenceResultHandler` and overwrite [handleSuccess](https://github.com/ngrx/platform/blob/master/modules/data/src/dataservices/persistence-result-handler.service.ts) method, the purpose is to parse the data got from `DataService` and retrieve the additional property then save it to `action.payload`. Here is a sample.
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
You need to create an `AdditionalPropertyPersistenceResultHandler` extends DefaultPersistenceResultHandler` and overwrite [handleSuccess](https://github.com/ngrx/platform/blob/master/modules/data/src/dataservices/persistence-result-handler.service.ts) method, the purpose is to parse the data got from `DataService` and retrieve the additional property then save it to `action.payload`. Here is a sample.
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 3]
```diff
@@ -220,6 +220,95 @@ The property values become the initial collection values for those properties wh
...
+
+
+```typescript
+export class AdditionalPropertyPersistenceResultHandler {
+  handleSuccess(originalAction: EntityAction): (data: any) => Action {
+    const actionHandler = super.handleSuccess(originalAction);
+    // here will return a factory to get a data handler to
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
    // return a factory to get a data handler to
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 4]
```diff
@@ -220,6 +220,95 @@ The property values become the initial collection values for those properties wh
...
+
+#### Step3: Register customized `EntityCollectionReducermethods` and `PersisntenceResultHandler`.
+Finally we need to register the `AdditionalPersistenceResultHandler` and `AdditionalEntityCollectionReducerMethods` to replace the default implementation.
+
+Register `AdditionalPersistenceResultHandler` in `NgModule`,
+
+```typescript
+@NgModule({
+  ...
+    { provide: PersistenceResultHandler, useClass: PagePersistenceResultHandler },
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
 { provide: PersistenceResultHandler, useClass: PagePersistenceResultHandler },
```

**Comment by JiaLiPassion** [comment]
@timdeschryver, thanks for the review, I have updated the doc!

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 5]
```diff
@@ -220,6 +220,96 @@ The property values become the initial collection values for those properties wh
...
+      }
+      return action;
+    };
+  }
+}
+```
+
+#### Step2: Overwrite `EntityCollectionReducerMethods` to Save additional property from action.payload to EntityCollection instance
+
+Now the additional property is set to `action.payload`, but we still need to set it to the instance of EntityCollection in `reducer`, to do that, we need to create a `AdditionalEntityCollectionReducerMethods extends EntityCollectionReducerMethods`, and overwrite the method match your `action`, for example, if the additional property `foo` only available in `getWithQuery action`, we can do like this.
```
[the remark this answers is not in the document]
Comment (timdeschryver) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: I'm not very familiar with ngrx/data and that's probably why I find the below confusing. 

>  and overwrite the method match your `action`, for example, if the additional property `foo` is only available in the `getWithQuery action`, we can do like this.


For example in the example below, I don't see a `getWithQuery action`

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 6]
```diff
@@ -220,6 +220,96 @@ The property values become the initial collection values for those properties wh
...
+```
+
+#### Step3: Register customized `EntityCollectionReducermethods` and `PersisntenceResultHandler`.
+Finally we need to register the `AdditionalPersistenceResultHandler` and `AdditionalEntityCollectionReducerMethods` to replace the default implementation.
+
+Register `AdditionalPersistenceResultHandler` in `NgModule`,
+
+```typescript
+@NgModule({
+  ...
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion

```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 7]
```diff
@@ -220,6 +220,96 @@ The property values become the initial collection values for those properties wh
...
+     return methodsClass.methods;
+  }
+}
+```
+
+then register this `AdditionalEntityCollectionReducerMethodsFactory` to `NgModule`,
+
+```typescript
+@NgModule({
+  ...
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion

```

File: projects/ngrx.io/content/guide/data/entity-metadata.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (JiaLiPassion) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: @timdeschryver, yeah, this is a little confusing, I will update the text.

**Comment by JiaLiPassion** [comment]
@timdeschryver, thank you for the detailed review, I updated the doc!

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 8]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
 
 The NgRx Data library generates selectors for these properties but has no way to update them. You'll have to create or extend the existing reducers to do that yourself.
 
+If the property you want to add comes from `backend`, you will need some additional work to make sure the property can be saved into the store from `Effects` correctly.
+
+#### Step1: Implement `PersistenceResultHandler` to save data from backend to action.payload
+
+You need to create an `AdditionalPersistenceResultHandler extends DefaultPersistenceResultHandler` and overwrite the [handleSuccess](https://github.com/ngrx/platform/blob/master/modules/data/src/dataservices/persistence-result-handler.service.ts) method, the purpose is to parse the data got from `DataService` and retrieve the additional property then save it to `action.payload`. Here is a sample.
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
Create a new class `AdditionalPersistenceResultHandler` that `extends DefaultPersistenceResultHandler` and overwrite the [handleSuccess](https://github.com/ngrx/platform/blob/master/modules/data/src/dataservices/persistence-result-handler.service.ts) method, the purpose is to parse the data received from the `DataService`, retrieve the additional property, and then save this to the `action.payload`. Here is a sample.
```

**Comment by JiaLiPassion** [comment]
@wesleygrimes, thanks, sorry, I didn't see the changes you requested, could you check it?

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 9]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
...
+
+```typescript
+export class AdditionalPropertyPersistenceResultHandler {
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]: Shouldn't this have `extends` ?
```suggestion
export class AdditionalPropertyPersistenceResultHandler extends DefaultPersistenceResultHandler {
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 10]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
...
+        // save the data.foo to action.payload.foo
+        (action as any).payload.foo = data.foo;
+      }
+      return action;
+    };
+  }
+}
+```
+
+#### Step2: Overwrite `EntityCollectionReducerMethods` to save additional property from action.payload to EntityCollection instance
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
#### Step2: Overwrite `EntityCollectionReducerMethods` to save the additional property from action.payload to the `EntityCollection` instance
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 11]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
...
+
+Now the additional property is set to `action.payload`, but we still need to set it to the instance of EntityCollection in the `reducer`, to do that, we need to create a `AdditionalEntityCollectionReducerMethods extends EntityCollectionReducerMethods`, and overwrite the method match your `action`, for example, if the additional property `foo` only available in `queryMany action(triggered by EntityCollectionService.getWithQuery)`, we can do like this.
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
Following the prior step, we have added the additional property to the `action.payload`. Up next we need to set it to the instance of EntityCollection in the `reducer`. In order to accomplish that, we need to create an `AdditionalEntityCollectionReducerMethods` that `extends EntityCollectionReducerMethods`. In addition, we will need to overwrite the method to match your `action`. For example, if the additional property `foo` is only available in `queryMany action(triggered by EntityCollectionService.getWithQuery)`, we can following this approach.
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 12]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
...
+   /** Create the  {EntityCollectionReducerMethods} for the named entity type */
+  create<T>(entityName: string): EntityCollectionReducerMethodMap<T> {
+    const definition = this.entityDefinitionService.getDefinition<T>(entityName);
+    const methodsClass = new AdditionalEntityCollectionReducerMethods(entityName, definition);
+     return methodsClass.methods;
+  }
+}
+```
+
+then register this `AdditionalEntityCollectionReducerMethodsFactory` to `NgModule`,
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
Register `AdditionalEntityCollectionReducerMethodsFactory` to `NgModule`,
```

Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
Register this `AdditionalEntityCollectionReducerMethodsFactory` to `NgModule`,
```

**Comment by wesleygrimes** [comment]
> @wesleygrimes, thanks, sorry, I didn't see the changes you requested, could you check it?

Sorry about that, they should be there now.

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 13]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
 
 The NgRx Data library generates selectors for these properties but has no way to update them. You'll have to create or extend the existing reducers to do that yourself.
 
+If the property you want to add comes from `backend`, you will need some additional work to make sure the property can be saved into the store from `Effects` correctly.
+
+#### Step1: Implement `PersistenceResultHandler` to save data from backend to action.payload
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
#### Step 1: Implement `PersistenceResultHandler` to save data from backend to action.payload
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md (hunk 10 above)
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
#### Step 2: Overwrite `EntityCollectionReducerMethods` to save additional property from action.payload to EntityCollection instance
```

**Comment by JiaLiPassion** [comment]
@wesleygrimes, thanks for the review, I have updated the doc.

**Comment by JiaLiPassion** [comment]
@wesleygrimes, sure, thanks, I have fixed those 3 `Steps`.

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 14]
```diff
@@ -220,6 +220,96 @@ The property values become the initial collection values for those properties wh
...
+
+You need to create an `AdditionalPersistenceResultHandler extends DefaultPersistenceResultHandler` and overwrite [handleSuccess](https://github.com/ngrx/platform/blob/master/modules/data/src/dataservices/persistence-result-handler.service.ts) method, the purpose is to parse the data got from `DataService` and retrieve the additional property then save it to `action.payload`. Here is a sample.
```
[the remark this answers is not in the document]
Comment (wesleygrimes) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: @JiaLiPassion I believe this still needs to be changed to `an example` not `a sample`.

File: projects/ngrx.io/content/guide/data/entity-metadata.md [hunk 15]
```diff
@@ -220,6 +220,92 @@ The property values become the initial collection values for those properties wh
...
+
+Register `AdditionalPersistenceResultHandler` in `NgModule`,
+
+```typescript
+@NgModule({
+  { provide: PersistenceResultHandler, useClass: PagePersistenceResultHandler },
+})
+```
+
+Register `AdditionalEntityCollectionReducerMethods`, to do that, we need to create a `AdditionalEntityCollectionReducerMethodFactory`, for details, see [Entity Reducer](guide/data/entity-reducer)
```
Comment (wesleygrimes) [inline projects/ngrx.io/content/guide/data/entity-metadata.md]:
```suggestion
Register `AdditionalEntityCollectionReducerMethods`, to do that, we need to create an `AdditionalEntityCollectionReducerMethodFactory`, for details, see [Entity Reducer](guide/data/entity-reducer)
```

File: projects/ngrx.io/content/guide/data/entity-metadata.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (JiaLiPassion) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: @wesleygrimes , sorry I missed this one, I just fixed it, thanks!

File: projects/ngrx.io/content/guide/data/entity-metadata.md (hunk 15 above)
Comment (JiaLiPassion) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: @wesleygrimes, thank you very much, I need to use some grammar check tool to do some check before commit the next time.

Comment (wesleygrimes) [reply projects/ngrx.io/content/guide/data/entity-metadata.md]: You're welcome. Thanks for contributing! We use Grammarly and find good success with it. It's free too!

**Comment by wesleygrimes** [comment]
LGTM

**Comment by JiaLiPassion** [comment]
@wesleygrimes, thanks, I will use `Grammarly`.

**Comment by AdditionAddict** [comment]
@JiaLiPassion I think this needs an update to step 1.

The default QUERY_MANY_SUCCESS reducer is expecting action.payload.data to be an array of entities but the QUERY_MANY action action.payload.data will be whatever the api returns.

Say it returns

interface QueryManyAPIResponse<T> {
    total: number,
    entities: T[]
}

You can add a property total to the action payload but action.payload.data needs to be the entities if you want the default reducer for a query many success to work.

```
@Injectable()
export class AdditionalPersistenceResultHandler  extends DefaultPersistenceResultHandler {

    handleSuccess(originalAction: EntityAction): (data: any) => Action {
        const actionHandler = super.handleSuccess(originalAction);

        return function(data: any) {
            /** Default success action */
            const successAction = actionHandler.call(this, data);

            /** Change payload for query many */
            if (successAction && data && data.total) {
                (successAction as any).payload.total = data.total;
            }
            if (successAction && data && data.entities) {
                (successAction as any).payload.data = data.entities;
            }

            return action;
        };
    }
}
```

**Pull request merged**